### PR TITLE
fix: duplicate CDK resource ID

### PR DESCRIPTION
**Pull request by WinterYukky**
#### Which issue(s) does this change fix?

#4034 
https://github.com/aws/aws-cdk/issues/21134

#### Why is this change necessary?
Currently, SAM CLI extract SamResourceId from last second string of `aws:cdk:path` value.
But, CDK possibly generate `aws:cdk:path` like next.

- Stack/aaa/Fn/Resource -> SamResouceId: Fn
- Stack/bbb/Fn/Resource -> SamResouceId: Fn
- Stack/ccc/Fn/Resource -> SamResouceId: Fn

In such cases, the SAM CLI considers these resources to have the same SamResouceId (i.e., Fn) and treats only the last lambda function found as an executable lambda function. In other words, it will generate lambda functions that are not found.

#### How does it address the issue?

See https://github.com/aws/aws-sam-cli/pull/4064#issuecomment-1187724150

#### Mandatory Checklist
**PRs will only be reviewed after checklist is complete**

- [x] Add input/output [type hints](https://docs.python.org/3/library/typing.html) to new functions/methods
- [ ] Write design document if needed ([Do I need to write a design document?](https://github.com/aws/aws-sam-cli/blob/develop/DEVELOPMENT_GUIDE.md#design-document))
- [x] Write/update unit tests
- [x] Write/update integration tests
- [ ] Write/update functional tests if needed
- [x] `make pr` passes
- [ ] `make update-reproducible-reqs` if dependencies were changed
- [ ] Write documentation

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0).

### CODE REVIEW

**Comment by jfuss** [comment]
Left a comment in https://github.com/aws/aws-sam-cli/issues/4034 but putting the important details here too for visibility with the team. 

We need to understand if this is the right approach. Combining the two fields seems reasonable but I need to do a deeper check in what we support now and why before I can understand the right approach (allowing the logical id and the cdk path, using "_" as a delimiter, etc).

Again thanks to @WinterYukky for catching this and @sinokuma for the initial report.

**Comment by WinterYukky** [comment]
Thanks @jfuss for checking the PR.
I was wondering why you use part of the cdk path (i.e Fn) for the ID value, but your explanation helped me understand.
This PR works by using underscores as delimiters, but may not be a good solution for users. It is better to use slashes as delimiters than to use underscores, at least.

Currently, SAM CLI find functions by `function_id` `name` `functionname`.
https://github.com/aws/aws-sam-cli/blob/develop/samcli/lib/providers/sam_function_provider.py#L110-L112

Consider the following example.
```json
{
  "Resources": {
    "aaaFn67B877B1": {
      "Type": "AWS::Lambda::Function",
      "Properties": {},
      "Metadata": {
        "aws:cdk:path": "Stack/aaa/Fn/Resource",
        "aws:asset:path": "/workspaces/cdk-sam-local/lib/handler",
        "aws:asset:is-bundled": false,
        "aws:asset:property": "Code"
      }
    },
    "bbbFn9F2D7876": {
      "Type": "AWS::Lambda::Function",
      "Properties": {},
      "Metadata": {
        "aws:cdk:path": "Stack/bbb/Fn/Resource",
        "aws:asset:path": "/workspaces/cdk-sam-local/lib/handler",
        "aws:asset:is-bundled": false,
        "aws:asset:property": "Code"
      }
    },
    "cccFn256E40E6": {
      "Type": "AWS::Lambda::Function",
      "Properties": {},
      "Metadata": {
        "aws:cdk:path": "Stack/ccc/Fn/Resource",
        "aws:asset:path": "/workspaces/cdk-sam-local/lib/handler",
        "aws:asset:is-bundled": false,
        "aws:asset:property": "Code"
      }
    },
    "CDKMetadata": {
      "Type": "AWS::CDK::Metadata",
      "Properties": {
        "Analytics": "v2:deflate64:H4sIAAAAAAAA/zWNXQqDMBCEz+J73FqFvtdCD2APIGvcyvqTgJu0DyF3b6IUFr7ZYZipoamhKvArpR6XcuUBwsuhXlSy+rDiNowI4emNdmyNerzNX0fFuEHo7ErZzoxKmh5FyAncM9IPofV6IdeikDpljp8qRnXk0uTEZjp6SKzfdeoydiSY5fK53iBdVczCXO7eON4IupM/SkZ1q8EAAAA="
      },
      "Metadata": {
        "aws:cdk:path": "Stack/CDKMetadata/Default"
      }
    }
  }
}
```

In this case, the SAM CLI resolves the functions into the following objects.

| LogicalId | function_id | name | functionname |
|-|-|-|-|
| aaaFn67B877B1 | Fn | aaaFn67B877B1 | aaaFn67B877B1 |
| bbbFn9F2D7876 | Fn | bbbFn9F2D7876 | bbbFn9F2D7876 |
| cccFn256E40E6 | Fn | cccFn256E40E6 | cccFn256E40E6 |

How about changing it as follows.

| LogicalId | function_id | name | functionname |
|-|-|-|-|
| aaaFn67B877B1 | aaaFn67B877B1 | Fn | aaaFn67B877B1 |
| bbbFn9F2D7876 | bbbFn9F2D7876 | Fn | bbbFn9F2D7876 |
| cccFn256E40E6 | cccFn256E40E6 | Fn | cccFn256E40E6 |

When change to the proposed solution, then I think SAM CLI allows use these lambda functions. When call by aaaFn67B877B1 (i.e LogicalId) then invoke the found function and when call by Fn then print warning log "Multiple functions found with keyword {name}! Function ... ".

**Comment by jfuss** [comment]
@WinterYukky I chatted with the team. Instead of concatenating parts of the `cdk:path`, we want to ensure that the LogicalId is callable. Concatenating, still requires the customer to understand how the `cdk:path` works and map it to their code. Instead, we feel that supporting the LogicalId should always work. We know this does not solve the how we handle the `cdk:path` but want to start from the LogicalId as a first step.

Would you be willing to update this PR to support this?

**Comment by WinterYukky** [comment]
@jfuss Of course! I'll try it.

**Review by WinterYukky — COMMENTED** [review]
@jfuss 
From the following results, the behavior appears to be good. Could you check the PR?

```bash
vscode ➜ /workspaces/aws-sam-cli (fix/duplicate-CDK-ID ✗) $ samdev local invoke Fn -t cdk.out/ExampleStack.template.json 

        SAM CLI now collects telemetry to better understand customer needs.

        You can OPT OUT and disable telemetry collection by setting the
        environment variable SAM_CLI_TELEMETRY=0 in your shell.
        Thanks for your help!

        Learn More: https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-sam-telemetry.html

Multiple functions found with keyword Fn! Function Example1FnCBC0C9FD will be invoked! If it's not the function you are going to invoke, please choose one of them from below:
Example1FnCBC0C9FD
Example2FnFBA58E3D
Invoking app.handler (nodejs16.x)
Skip pulling image and use local one: public.ecr.aws/sam/emulation-nodejs16.x:rapid-1.53.0-x86_64.

Mounting /workspaces/aws-sam-cli as /var/task:ro,delegated inside runtime container
START RequestId: a5b9b93a-d047-4901-924c-4405e7fb338a Version: $LATEST
{"statusCode":200,"body":"{\"message\":\"Hello from Lambda!\"}"}END RequestId: a5b9b93a-d047-4901-924c-4405e7fb338a
REPORT RequestId: a5b9b93a-d047-4901-924c-4405e7fb338a  Init Duration: 0.20 ms  Duration: 219.29 ms   Billed Duration: 220 ms  Memory Size: 128 MB     Max Memory Used: 128 MB
```

<details><summary>Used template</summary><div>

```json
{
    "Resources": {
     "Example1FnServiceRole6C38A161": {
      "Type": "AWS::IAM::Role",
      "Properties": {
       "AssumeRolePolicyDocument": {
        "Statement": [
         {
          "Action": "sts:AssumeRole",
          "Effect": "Allow",
          "Principal": {
           "Service": "lambda.amazonaws.com"
          }
         }
        ],
        "Version": "2012-10-17"
       },
       "ManagedPolicyArns": [
        {
         "Fn::Join": [
          "",
          [
           "arn:",
           {
            "Ref": "AWS::Partition"
           },
           ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
          ]
         ]
        }
       ]
      },
      "Metadata": {
       "aws:cdk:path": "ExampleStack/Example1/Fn/ServiceRole/Resource"
      }
     },
     "Example1FnCBC0C9FD": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
       "Code": {
        "S3Bucket": {
         "Fn::Sub": "cdk-hnb659fds-assets-${AWS::AccountId}-us-east-1"
        },
        "S3Key": "448f44cb8523ba7e460e321e56bc18df73372207e9166e1089b2ef72cbe168c8.zip"
       },
       "Role": {
        "Fn::GetAtt": [
         "Example1FnServiceRole6C38A161",
         "Arn"
        ]
       },
       "Handler": "app.handler",
       "Runtime": "nodejs16.x"
      },
      "DependsOn": [
       "Example1FnServiceRole6C38A161"
      ],
      "Metadata": {
       "aws:cdk:path": "ExampleStack/Example1/Fn/Resource",
       "aws:asset:path": "/workspaces/aws-sam-cli",
       "aws:asset:is-bundled": false,
       "aws:asset:property": "Code"
      }
     },
     "Example2FnServiceRole839DD1B5": {
      "Type": "AWS::IAM::Role",
      "Properties": {
       "AssumeRolePolicyDocument": {
        "Statement": [
         {
          "Action": "sts:AssumeRole",
          "Effect": "Allow",
          "Principal": {
           "Service": "lambda.amazonaws.com"
          }
         }
        ],
        "Version": "2012-10-17"
       },
       "ManagedPolicyArns": [
        {
         "Fn::Join": [
          "",
          [
           "arn:",
           {
            "Ref": "AWS::Partition"
           },
           ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole"
          ]
         ]
        }
       ]
      },
      "Metadata": {
       "aws:cdk:path": "ExampleStack/Example2/Fn/ServiceRole/Resource"
      }
     },
     "Example2FnFBA58E3D": {
      "Type": "AWS::Lambda::Function",
      "Properties": {
       "Code": {
        "S3Bucket": {
         "Fn::Sub": "cdk-hnb659fds-assets-${AWS::AccountId}-us-east-1"
        },
        "S3Key": "448f44cb8523ba7e460e321e56bc18df73372207e9166e1089b2ef72cbe168c8.zip"
       },
       "Role": {
        "Fn::GetAtt": [
         "Example2FnServiceRole839DD1B5",
         "Arn"
        ]
       },
       "Handler": "app.handler",
       "Runtime": "nodejs16.x"
      },
      "DependsOn": [
       "Example2FnServiceRole839DD1B5"
      ],
      "Metadata": {
       "aws:cdk:path": "ExampleStack/Example2/Fn/Resource",
       "aws:asset:path": "/workspaces/aws-sam-cli",
       "aws:asset:is-bundled": false,
       "aws:asset:property": "Code"
      }
     },
     "CDKMetadata": {
      "Type": "AWS::CDK::Metadata",
      "Properties": {
       "Analytics": "v2:deflate64:H4sIAAAAAAAA/zWM0QqDMAxFv8X3mjmFvc/BPkA/QGLtJLamYNrtofjvq45B4CTnkltDU0NV4EdKPdnS0QipD6itympIDtdxQkjPyDqQZ/V48X/fFeEKqfPOHPrgrqQZUMQEgfuBfEMbtTWhRcnxKXP/TDyfT0Z83HRO2E8GFrm8rzfIUxWLEJVb5ECrge7HL6MfX3uuAAAA"
      },
      "Metadata": {
       "aws:cdk:path": "ExampleStack/CDKMetadata/Default"
      }
     }
    },
    "Parameters": {
     "BootstrapVersion": {
      "Type": "AWS::SSM::Parameter::Value<String>",
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]"
     }
    },
    "Rules": {
     "CheckBootstrapVersion": {
      "Assertions": [
       {
        "Assert": {
         "Fn::Not": [
          {
           "Fn::Contains": [
            [
             "1",
             "2",
             "3",
             "4",
             "5"
            ],
            {
             "Ref": "BootstrapVersion"
            }
           ]
          }
         ]
        },
        "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI."
       }
      ]
     }
    }
   }
```

</div></details>

**Comment by jfuss** [comment]
@WinterYukky I should be able to get you a review this week. Sorry for the delay.

File: samcli/lib/providers/provider.py
Comment (jfuss) [inline samcli/lib/providers/provider.py:896]: I would prefer to find ways to leave the default to be `resource.name`. This is a much more maintainable solution. For example, if we add a new builder this will try to access `resource.layer_id` instead of "just working". If we need to change this, I am ok but let's change this to check for each resource and access the attribute we need. Otherwise default to `resource.name`.

File: samcli/lib/providers/sam_layer_provider.py
Comment (jfuss) [inline samcli/lib/providers/sam_layer_provider.py:167]: I am not exactly understanding what layer_id is meant to represent. It looks like it is just the `layer_logical_id` but we have that within the `arn` property. So it's not clear why we need another property for the same thing.

File: samcli/lib/providers/sam_layer_provider.py
Comment (jfuss) [inline samcli/lib/providers/sam_layer_provider.py:157]: I think I will need more time to digest this. But it's not super clear why this is here, how this changes part of the code base for getting the name or logical id for this resource.

**Review by jfuss — CHANGES_REQUESTED** [review]
I left a couple comments but think I need to do a fresh pass when my brain is fresh. Just having a hard time following how `name` and `layer_id` fit into everything. Ideally, we could follow the pattern we have for functions but maybe there is a reason they need to be different that I don't quite get yet. 

I at least wanted to get some feedback out to you now.

**Comment by WinterYukky** [comment]
> I left a couple comments but think I need to do a fresh pass when my brain is fresh. Just having a hard time following how `name` and `layer_id` fit into everything. Ideally, we could follow the pattern we have for functions but maybe there is a reason they need to be different that I don't quite get yet.
> 
> I at least wanted to get some feedback out to you now.

Thanks for reviewing @jfuss!
I recheck the my change and will reply.

File: samcli/lib/providers/provider.py
Comment (WinterYukky) [reply samcli/lib/providers/provider.py:744]: Yes, it is similar to the `get_full_path` implementation.

File: samcli/lib/providers/provider.py
Comment (WinterYukky) [reply samcli/lib/providers/provider.py:896]: You are correct.
This time, we noticed that the value of resource.name coming from the CDK output template has been changed, and the name of the generated build directory has changed. Therefore, we have included this workaround for backward compatibility. If you are not affected by changing the name of the build directory, you can use the resource.name as it is. If there is no guarantee of this, then resource.name is the default.

File: samcli/lib/providers/sam_layer_provider.py
Comment (WinterYukky) [reply samcli/lib/providers/sam_layer_provider.py:167]: Like `Function`, `Layer` also used potentially duplicate values (e.g. Fn) extracted from `aws:cdk:path` for `arn`. Thus, the data structure was as follows, which could cause the same problem as with `Function`.

| LogicalId | arn | layer_id(SamResourceId) | name |
|-|-|-|-|
| aaaLayer123 | Layer | Layer | Layer |
| bbbLayer456 | Layer | Layer | Layer |
| cccLayer789 | Layer | Layer | Layer |

To deal with the same problem as `Function`, it was necessary to change the data structure to the following one.

| LogicalId | arn | layer_id(SamResourceId) | name |
|-|-|-|-|
| aaaLayer123 | aaaLayer123 | aaaLayer123 | Layer |
| bbbLayer456 | bbbLayer456 | bbbLayer456 | Layer |
| cccLayer789 | cccLayer789 | cccLayer789 | Layer |

However, the `layer_id` and `name` resolution logic was structured in the following order of reference.

```mermaid
flowchart LR
    layer_id -- refer if not --> name -- refer if not --> arn
```

Therefore, changing `arn` to a `logical_id` would change `name`, and specifying `name` would next change `layer_id`, which references `name`, so there was a need to specify each completely.

I think this change is necessary to have backward compatibility because it supported potentially duplicate values (e.g. Fn) extracted from `aws:cdk:path` as search conditions for `Layer`.

File: samcli/lib/providers/sam_layer_provider.py
Comment (WinterYukky) [reply samcli/lib/providers/sam_layer_provider.py]: The code was indeed a bit difficult to read the intent. This is a change that applies only to CDK resources, and requires that `layer_id` and `name` be specified when initializing the `LayerVersion` as described at https://github.com/aws/aws-sam-cli/pull/4064#discussion_r939480604 This is a change that applies only to CDK resources.
I have refactored it so that it is clear that it applies only to CDK resources.

Comment (jfuss) [reply samcli/lib/providers/sam_layer_provider.py]: Maybe layer is different but it might be easier to understand and read if we don't do conditional things per IAC. This can make expanding and supporting Terraform or other IAC much more difficult in the future.

File: samcli/lib/providers/sam_layer_provider.py
Comment (jfuss) [reply samcli/lib/providers/sam_layer_provider.py]: So if I am understanding what you are saying here is that they way we get the information from the `aws:cdk:path` and apply it into our object is the root of the issue here because we override properties in the generated template from CDK.

Is that a good summary?

**Review by jfuss — CHANGES_REQUESTED** [review]
Still processing the changes but overall I think this looks good. 

Would you be able to add integration tests for this? And cover the cases like 1) invoking with logical id if two have similar cdk paths 2) invoke the using the "cdk id" (for backwards compatibility) 3) If there are any others you can think through.

Integration tests will really help me in understanding this is working as we intended.

File: samcli/lib/providers/sam_layer_provider.py
Comment (WinterYukky) [reply samcli/lib/providers/sam_layer_provider.py]: Yes, the cdk id must be set to name and the logical id to layer_id, and this is the escape hatch that makes it happen.

**Review by jfuss — APPROVED** [review]
Seems good to me. Thanks for handling all the feedback.

We do require two approvals before merging. So I will find another person to review. Hopefully, we got most of the things handled through my review.

Thanks again for the contribution!

File: samcli/lib/providers/provider.py
Comment (moelasmar) [inline samcli/lib/providers/provider.py:107]: I think this function will mix between the CDK ID, and logical ID. It should return the same value as full_path in CFN use cases, but in CDK use case it will return the <stack CDK ID>/<Function Logical ID>. I am not sure how this mix will help the customer to identify a lambda function.

File: samcli/lib/providers/provider.py
Comment (moelasmar) [inline samcli/lib/providers/provider.py:393]: same as `full_name` in Function

File: samcli/lib/providers/sam_function_provider.py
Comment (moelasmar) [reply samcli/lib/providers/sam_function_provider.py:112]: why do not you use the `f.full_path` as well? The customer should be able to use `<stack cdk id>/<function cdk id>` as an identifier for a function.

File: samcli/lib/samlib/resource_metadata_normalizer.py
Comment (moelasmar) [inline samcli/lib/samlib/resource_metadata_normalizer.py:248]: if we add this change, the resource name is something related to CDK only. This way will break the Other IaC platforms support. 

The idea is we either define a lambda function using the logical Id or the custom defined Id (IaC Id) using `SamResourceId` metadata. Only in case of CDK, as the `cdk synth` command does not add the `SamSourceId` metadata, we use the `aws:cdk:path` to be the custom defined id (IaC Id).

**Comment by moelasmar** [comment]
Thanks @WinterYukky for your effort. I have some concern about the solution you suggested to fix this problem. My concern is it could not be extend to other IaC frameworks. The `full_name` property you added to the lambda functions, and layers seems is only applicable for CDK use case, and this breaks our current implementation to support other IaC frameworks like Terraform.

As per [Jacob's comment](https://github.com/aws/aws-sam-cli/issues/4034#issuecomment-1187347966) in the opened issue. The solution we are looking for is to enable the customer to use logical id for search, and not to fix the CDK id extraction.So, I would suggest to define two ids for Functions, Layers, and Stacks (Logical Id, and Custom Id), and to have `full_logical_id` which will be `<parent stack full_logical_id>/<resource logical id>` and `full_custom_id` which will be `<parent stack full_custom_id>/<resource custom id>`. So, the customer can use either the a path of logical ids or a path of custom ids to define a resource, as I do not think that it is a valid use case for the customer to mix between the 2 id types to define a resource.

I also suggest to keep `ResourceMetadataNormalizer.get_resource_id` as it is, and remove `ResourceMetadataNormalizer.get_resource_name`. As, the custom logic to determine the resource custom id (CDK id) is only applicable for CDK use case, and this id is always equivalent to the custom id. The best solution here is to add a change to CDK so during synthesize step, it adds the `SamResourceId` metadata, and set its value to the resource CDK Id.

**Comment by WinterYukky** [comment]
Thanks for reviewing @moelasmar!
Your propose that using new properties `custom_id` and `full_custom_id` looks like nice. I also understand your thoughts on other IaC frameworks. 

However, I don't agree to keep `ResourceMetadataNormalizer.get_resource_id` as it is. Because CDK ID (which SAM expects) can be duplicates. The solution can't fix this bug.

So I thought I should use the `IaCPluginInterface` to implement per-IaC behavior, but I don't know if that is appropriate. Can you help me?

File: samcli/lib/providers/provider.py
Comment (WinterYukky) [reply samcli/lib/providers/provider.py:107]: You are correct.
I will fix to using only CDK ID.

**Comment by moelasmar** [comment]
Sorry @WinterYukky for my slow reply. I want to leave the `ResourceMetadataNormalizer.get_resource_id` for now as in my opinion we should fix it from the CDK side, we need to push a PR in the CDK repo to add a new Metadata property to get the resource CDK id. Any fix we are going to do in SAM CLI is a workaround in my opinion, that is why I am thinking to leave it as is right now, till we can do the CDK change.

For this PR, we can change its purpose to enable the customers to use the Logical Ids to invoke a lambda function in case if there is any issue with using the CDK ID, so the customers can unblock themselves.

**Comment by mndeveci** [comment]
Thanks for your contribution @WinterYukky ! I will be converting this PR into draft until you will have some time to address comments from second part of @moelasmar's suggestion.

**Comment by WinterYukky** [comment]
@mndeveci 
I'm sorry. I'm busy now but working this issue slowly. I'm never abandonment.
When I'm ready let you know.

**Comment by jfuss** [comment]
Closing as this appears stale. If you are still interesting in contribution this, feel free to reopen or create a new PR.

**Pull request closed without merging**
